#!/bin/bash
#PBS -j oe
#PBS -m ae
#PBS -N MitoFinder_PMC
#PBS -l select=1:ncpus=20:mem=20gb
#PBS -l walltime=5:00:00
#PBS -M you.email@uni.com
#PBS -J 0-67

cd $PBS_O_WORKDIR
ncpu=$(qstat -f $PBS_JOBID | grep "Resource_List.ncpus" | cut -d= -f2 | sed 's/ //g') 
s=$(if [ $ncpu == 1 ]; then echo ""; else echo "s"; fi)
mem=$(qstat -f $PBS_JOBID | grep "Resource_List.mem" | cut -d= -f2 | sed 's/ //g')
echo $(date)
echo "------------------------------------------------------"
echo "This job is allocated "$ncpu" CPU core$s and $mem on " cat $PBS_NODEFILE | uniq
echo "------------------------------------------------------"
echo "PBS: Submitted to $PBS_QUEUE@$PBS_O_HOST"
echo "PBS: Working directory is $PBS_O_WORKDIR"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"
echo "------------------------------------------------------"

# Make sure required modules are discoverable and load them
shopt -s expand_aliases
source /etc/profile.d/modules.sh

module load java/1.8.0
module load blast/2.12.0
module load spades/3.15.5
#source activate megahit
# source activate spades-3.15.2

############### MITOFINDER RUN 1 - ASSEMBLE & ANNOTATE MT CONTIGS ##############
                #---- COMMENT OUT FOR RUN2 ----#

mkdir /scratch/MitoFinder/MF1_out
cd /scratch/MitoFinder/MF1_out
# get array index
id=${PBS_ARRAY_INDEX}
# Get list of species
indirs=($(ls /home/MitoFinder/2_cleanedreads_PMC))
# Get infile
sample="${indirs[$id]}"

# make output file
#sample=$(echo $indir | cut -d/ -f2)
r1=$(echo "/home/MitoFinder/2_cleanedreads_PMC/$sample/split-adapter-quality-trimmed/$sample-READ1.fastq.gz")
r2=$(echo "/home/MitoFinder/2_cleanedreads_PMC/$sample/split-adapter-quality-trimmed/$sample-READ2.fastq.gz")

# run mitofinder to assemble with metaspades
# If you are using capture data (e.g. UCE), consider specifying the --min-contig-size parameter. The default value is 1,000 bp which may be too high for mitochondrial contigs assembled from off-target reads. The same applies for the parameter --blast-size (default: 30%). From fasta.summary obtained in Phyluce, min contig size is 140bp.
/home/MitoFinder/MitoFinderPackage/mitofinder --metaspades \
        -j $sample \
        -1 $r1 \
        -2 $r2 \
        -r RefSeq_mtDNA_Pomacanthidae.gb \
        -o 2 \
        -p 20 \
        -t mitfi \
        -e 0.000001 \
        --min-contig-size 500 \
        --max-memory 100 \
        --new-genes 

                #------------------------------#

# Cat all the contigs from the first MitoFinder run for each species - see annotated script for more details if needed

############### MITOFINDER RUN 2 - ALLOW INTRONS ##############
# mkdir /scratch/MitoFinder/MF2_out
# cd /scratch/MitoFinder/MF2_out
# get array index
# id=${PBS_ARRAY_INDEX}
# # get list of contig files from first MF run
# indirs=($(ls /scratch/MitoFinder/PMC_contigs_MFRUN1))
# sample="${indirs[$id]}"
# contigs="/scratch/MitoFinder/PMC_contigs_MFRUN1/$sample/${sample}_MFRUN1_mtDNA_contigs.fasta"
# # run mitofinder to search for genes with introns
# mitofinder -j  $sample \
#         -a $contigs
#         -r PMC_mtRefSeq.gb \
#         -o 2 \
#         -p 20 \
#         -m 40 \
#         -t mitfi \
#         --new-genes \
#         --allow-intron \
#         --intron-size 12000 \
#         --cds-merge \
#         --adjust-direction
